tinymix.c

```diff
@@ -196,27 +197,45 @@ static void tinymix_set_value(struct mixer *mixer, const char *control,
     type = mixer_ctl_get_type(ctl);
     num_ctl_values = mixer_ctl_get_num_values(ctl);
 
-    if (isdigit(values[0][0])) {
-        if (num_values == 1) {
-            /* Set all values the same */
-            int value = atoi(values[0]);
+    if ((isdigit(values[0][0])) || (values[0][0] == '-')) {
+        char *endptr = NULL;
+        long value;
+        /* Initialize a function pointer of mixer_ctl_set_... type */
+        int (*mixer_ctl_set)(struct mixer_ctl*, unsigned int, int) = NULL;
+        errno = 0;
 
-            for (i = 0; i < num_ctl_values; i++) {
-                if (mixer_ctl_set_value(ctl, i, value)) {
-                    fprintf(stderr, "Error: invalid value\n");
-                    return;
-                }
-            }
-        } else {
-            /* Set multiple values */
-            if (num_values > num_ctl_values) {
-                fprintf(stderr,
-                        "Error: %d values given, but control only takes %d\n",
-                        num_values, num_ctl_values);
+        if (num_values > num_ctl_values) {
+            fprintf(stderr,
+                    "Error: %d values given, but control only takes %d\n",
+                    num_values, num_ctl_values);
+            return;
+        }
+
+        for (i = 0; i < num_values; i++) {
+            value = strtol(values[i], &endptr, 10);
+
+            if (errno != 0) {
+                perror("strtol");
```

Comment:
Since you're using strtol and error-checking, it would be nice to also check the case of unintelligible input where (endptr==values[i]), no recognisable digits at all, and (*endptr!='\0' && *endptr!='%') a non-digit charater was found